docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+    import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;
 
 Then run the compiler:
 
 .. code-block:: bash
 
-  solc github.com/ethereum/dapp-bin/=/usr/local/dapp-bin/ source.sol
+    solc github.com/ethereum/dapp-bin/=/usr/local/dapp-bin/ source.sol
+
+As a more complex example, suppose you rely on a module that uses an old version of dapp-bin that
+you checked out to ``/usr/local/dapp-bin_old``, then you can run:
+
+.. code-block:: bash
+
+    solc module1:github.com/ethereum/dapp-bin/=/usr/local/dapp-bin/ \
+         module2:github.com/ethereum/dapp-bin/=/usr/local/dapp-bin_old/ \
+         source.sol
+
+This means that all imports in ``module2`` point to the old version but imports in ``module1``
+point to the new version.
+
+Here are the detailed rules governing the behavior of remappings:
+
+#. **Remappings only affect the translation between import paths and import keys.**
+
+   Import keys added via other means cannot be remapped.
+   For example the paths you specify on the command-line and the ones in ``urls`` in Standard JSON
+   are not affected.
+
+    .. code-block:: bash
+
+        solc /project=/contracts /project/contract.sol # Import key: /project/contract.sol
+
+#. **Context and prefix must match import keys, not import paths.**
+
+   - This means that you cannot remap ``./`` or ``../`` directly since they are replaced during
+     translation to import keys but you can remap the source locations they resolve into:
+
+     .. code-block:: bash
+
+         solc ./=a /project=b /project/contract.sol
+
+     .. code-block:: solidity
+         :caption: /project/contract.sol
+
+         import "./util.sol" as util; // Import key: b/util.sol
+
+   - You cannot remap base path or any other part of the path that is only added when the file is
+     looked up on the underlying filesystem.
+
+     .. code-block:: bash
+
+         solc /project=/contracts /project/contract.sol --base-path /project
+
+     .. code-block:: solidity
+         :caption: /project/contract.sol
+
+         import "util.sol" as util; // Import key: util.sol
 
-As a more complex example, suppose you rely on a module that uses an old
-version of dapp-bin that you checked out to ``/usr/local/dapp-bin_old``, then you can run:
+#. **Target is inserted directly into the import key and does not necessarily have to be a valid path.**
+
+   - It can be anything as long as the file loader can handle it.
+     In case of the command-line interface this includes also relative paths.
+     When using the JavaScript interface you can just as well use URLs and abstract identifiers if
+     your callback can handle them.
+
+   - Remapping happens after paths relative to the source directory have already been resolved.
+     This means that targets starting with ``./`` and ``../`` have no special meaning and are
+     relative to the base directory rather than to the source location.
```

Comment:
I wonder if might be better to disallow remapping targets to start with `./` or `../`. I think they are not useful in practice and they are ambiguous (I originally assumed they would actually work like imports relative to source but they do not).